installation problem

Hello, 
I am really interested in using your program, but I keep getting an error when trying the example script.
The first error is in this line:
carm = deepdrr.CArm(isocenter=center)
I get this error : TypeError: __init__() missing 1 required positional argument: 'isocenter_distance'

So I add a random value of 800 to read : carm = deepdrr.CArm(isocenter=center, isocenter_distance=800)

and it runs. But in this line a get a long error message:
CompileError: nvcc compilation of C:\Users\Samer\AppData\Local\Temp\tmprdw5qsft\kernel.cu failed
[command: nvcc --cubin -D NUM_MATERIALS=3 -arch sm_75 -m64 -IC:\Users\Samer\anaconda3\envs\deepdrr2\Lib\site-packages\deepdrr\projector\cubic -Ic:\users\samer\anaconda3\envs\deepdrr2\lib\site-packages\pycuda\cuda kernel.cu]
[stdout:
kernel.cu(351): error: expected an expression

kernel.cu(363): error: expected an expression

kernel.cu(371): error: expected an expression

kernel.cu(324): warning: variable "px" was set but never used

kernel.cu(324): warning: variable "py" was set but never used

kernel.cu(324): warning: variable "pz" was set but never used

kernel.cu(327): warning: variable "boundary_factor" was set but never used

3 errors detected in the compilation of "kernel.cu".
kernel.cu
]



Some Information about my system. I am running python 3.9 under windows 10 through a jupyter Notebook.
I have CUDA 11 installed. I tried using pycuda (https://documen.tician.de/pycuda/tutorial.html) And I had no errors.
I can run tensorflow CNN on my machine without errors edit: (in R).

edit: I also added the path to cl.exe to the envoronment variables.

I would be very thankful if you could you maybe help find where this error is coming from.

Hello,
First I apologise for the long response times. I am trying this on my home computer, so I can only reponse or troubleshoot in the evinings. 

@mathiasunberath About verbosity I owe you an apology. THe error message in Jupyter notebook is longer, but I thought the line numbers would be enough. I will copy the complete error message at the end of this message.

As you may have guessed I am a relative noob. So I will write down what I tried:
I tried compiling the kernel.cu file in the temp directory"nvcc compilation of C:\Users\Samer\AppData\Local\Temp\tmp4lbohdb3\kernel.cu". 
But I got te tresponse: "fatal error C1083: Cannot open include file: 'cubicTex3D.cu': No such file or directory". 
So I moved the kernel.cu file into this folder"C:\Users\Samer\anaconda3\envs\deepdrr2\Lib\site-packages\deepdrr\projector\cubic" and tried compiling again and again I got the same response: No such file or directory

I will try further tonight and come back to you if I get any further.



@benjamindkilleen 

I installed gcc-core, gcc debug info, gcc fortran, gcc g++, gcc objc and gcc objc ++ version 9.3.0-2 using cygwin installer. Install location is in c: and the c:\cygwin64\bin is added to the user path. I tested the gcc version from the command prompt and I got a reponse. 


Thank you for your assistance.



The complete error message (of the last step of the example script):
---------------------------------------------------------------------------
CompileError                              Traceback (most recent call last)
<ipython-input-6-45e9baba820c> in <module>
----> 1 with deepdrr.Projector(volume, camera_intrinsics, carm) as projector:
      2   projection = projector()

~\anaconda3\envs\deepdrr2\lib\site-packages\deepdrr\projector\projector.py in __init__(self, volume, camera_intrinsics, carm, step, mode, spectrum, add_scatter, add_noise, photon_count, threads, max_block_index, collected_energy, neglog)
    106 
    107         # compile the module
--> 108         self.mod = _get_kernel_projector_module(self.num_materials) # TODO: make this not a compile-time option.
    109         self.project_kernel = self.mod.get_function("projectKernel")
    110 

~\anaconda3\envs\deepdrr2\lib\site-packages\deepdrr\projector\projector.py in _get_kernel_projector_module(num_materials)
     43 
     44     # TODO: replace the NUM_MATERIALS junk with some elegant meta-programming.
---> 45     return SourceModule(source, include_dirs=[bicubic_path], no_extern_c=True, options=['-D', f'NUM_MATERIALS={num_materials}'])
     46 
     47 

~\anaconda3\envs\deepdrr2\lib\site-packages\pycuda\compiler.py in __init__(self, source, nvcc, options, keep, no_extern_c, arch, code, cache_dir, include_dirs)
    288         self._check_arch(arch)
    289 
--> 290         cubin = compile(source, nvcc, options, keep, no_extern_c,
    291                 arch, code, cache_dir, include_dirs)
    292 

~\anaconda3\envs\deepdrr2\lib\site-packages\pycuda\compiler.py in compile(source, nvcc, options, keep, no_extern_c, arch, code, cache_dir, include_dirs, target)
    252         options.append("-I"+i)
    253 
--> 254     return compile_plain(source, options, keep, nvcc, cache_dir, target)
    255 
    256 

~\anaconda3\envs\deepdrr2\lib\site-packages\pycuda\compiler.py in compile_plain(source, options, keep, nvcc, cache_dir, target)
    133                     "encountered an error")
    134         from pycuda.driver import CompileError
--> 135         raise CompileError("nvcc compilation of %s failed" % cu_file_path,
    136                 cmdline, stdout=stdout.decode("utf-8", "replace"),
    137                 stderr=stderr.decode("utf-8", "replace"))

CompileError: nvcc compilation of C:\Users\Samer\AppData\Local\Temp\tmpyoq75wm0\kernel.cu failed
[command: nvcc --cubin -D NUM_MATERIALS=3 -arch sm_75 -m64 -IC:\Users\Samer\anaconda3\envs\deepdrr2\Lib\site-packages\deepdrr\projector\cubic -Ic:\users\samer\anaconda3\envs\deepdrr2\lib\site-packages\pycuda\cuda kernel.cu]
[stdout:
kernel.cu(351): error: expected an expression

kernel.cu(363): error: expected an expression

kernel.cu(371): error: expected an expression

kernel.cu(324): warning: variable "px" was set but never used

kernel.cu(324): warning: variable "py" was set but never used

kernel.cu(324): warning: variable "pz" was set but never used

kernel.cu(327): warning: variable "boundary_factor" was set but never used

3 errors detected in the compilation of "kernel.cu".
kernel.cu
]

--- a/deepdrr/projector/project_kernel.cu
+++ b/deepdrr/projector/project_kernel.cu
@@ -52,59 +52,59 @@ texture<float, 3, cudaReadModeElementType> seg(12);
 texture<float, 3, cudaReadModeElementType> seg(13);
 #endif
 
-#define UPDATE(multiplier, n) ({\
+#define UPDATE(multiplier, n) {\
     output[idx + (n)] += (multiplier) * tex3D(volume, px, py, pz) * round(cubicTex3D(seg(n), px, py, pz));\
-})
+}
 
 #if NUM_MATERIALS == 1
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
-})
+}
 #elif NUM_MATERIALS == 2
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
-})
+}
 #elif NUM_MATERIALS == 3
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
-})
+}
 #elif NUM_MATERIALS == 4
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
     UPDATE(multiplier, 3);\
-})
+}
 #elif NUM_MATERIALS == 5
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
     UPDATE(multiplier, 3);\
     UPDATE(multiplier, 4);\
-})
+}
 #elif NUM_MATERIALS == 6
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
     UPDATE(multiplier, 4);\
     UPDATE(multiplier, 5);\
-})  
+}
 #elif NUM_MATERIALS == 7
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
     UPDATE(multiplier, 4);\
     UPDATE(multiplier, 5);\
     UPDATE(multiplier, 6);\
-})
+}
 #elif NUM_MATERIALS == 8
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -112,9 +112,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 5);\
     UPDATE(multiplier, 6);\
     UPDATE(multiplier, 7);\
-})
+}
 #elif NUM_MATERIALS == 9
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -123,9 +123,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 6);\
     UPDATE(multiplier, 7);\
     UPDATE(multiplier, 8);\
-})
+}
 #elif NUM_MATERIALS == 10
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -135,9 +135,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 7);\
     UPDATE(multiplier, 8);\
     UPDATE(multiplier, 9);\
-})
+}
 #elif NUM_MATERIALS == 11
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -148,9 +148,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 8);\
     UPDATE(multiplier, 9);\
     UPDATE(multiplierl, 10);\
-})
+}
 #elif NUM_MATERIALS == 12
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -162,9 +162,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 9);\
     UPDATE(multiplier, 10);\
     UPDATE(multiplier, 11);\
-})
+}
 #elif NUM_MATERIALS == 13
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -177,9 +177,9 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 10);\
     UPDATE(multiplier, 11);\
     UPDATE(multiplier, 12);\
-})
+}
 #elif NUM_MATERIALS == 14
-#define INTERPOLATE(multiplier) ({\
+#define INTERPOLATE(multiplier) {\
     UPDATE(multiplier, 0);\
     UPDATE(multiplier, 1);\
     UPDATE(multiplier, 2);\
@@ -193,11 +193,11 @@ texture<float, 3, cudaReadModeElementType> seg(13);
     UPDATE(multiplier, 11);\
     UPDATE(multiplier, 12);\
     UPDATE(multiplier, 13);\
-})
+}
 #else
 #define INTERPOLATE(multiplier) {\
     fprintf(stderr, "NUM_MATERIALS not in [1, 14]");\
-)
+}
 #endif
 
 // the CT volume (used to be tex_density)

